### DISCUSSION TITLE
How to speed up response from RFEM

### DISCUSSION THREAD

**Discussion by BurakYyurt** [Dlubal-Software/RFEM_Python_Client#220] [Q&A] [answered]
Hello all,

I am attaching 2 scripts with a simple rfem6 file. 

those scripts does following things:
create_loads: 
connects to the existing model, 
creates 135 load cases in it, and applies a member load to a member set that is already in the model for each newly created load case.

node_reactions: 
connects to the existing model, 
goes through the load cases, reading specific node reactions and spits out to the screen. Later I will use these numbers else where but for the time being I am OK with print screen.

Now I have some questions:

1) Each time create_load creates a load case & adds a member load, it takes up to 5 seconds. The gui freezes and "application locked" warning appears on the bottom left. It would not be too much of a problem but I have 135 of cases. The whole reason I am using an API is to speed up the process. It takes too long to create and apply loads. 

2) Each time it spits out a result from nodes, does the same "locked" thing which takes up to 1-2 seconds. Maybe creating load case no problem, but for getting a result for each node for each load case it becomes very long.

So for the first 2 questions: is it possible to make the process smoother (or faster)? or maybe I am doing something fundamentally wrong, so any help would be much appreciated.

3) In node_reactions script, I had to recreate the list of load cases, which I think is not the best way to do it. So Is there a way for me to iterate through the existing load cases of the currently connected model ?

Although my questions are very specific, my intention is to get a grasp of the logic. I have some experience with both python and rfem but this is the first time i use an API. 

Thanks in advance! 

[beams.zip](https://github.com/Dlubal-Software/RFEM_Python_Client/files/10705160/beams.zip)

create_loads:

```python
from RFEM.initModel import *
from RFEM.LoadCasesAndCombinations.loadCase import *
from RFEM.Loads.membersetload import *


Model(False, "beams.rf6")

# Create Load Cases
n_cases = 135
case_names = ["NL1-P" + str(i) for i in range(1, n_cases + 1)]

increment = 0.100
distance_a = 0
distance_b = 0.620
distance_c = 1.480
force = 184000

for i, name in enumerate(case_names):
    LoadCase(i + 1, name, [False])
    MemberSetLoad.Force(1, i+1, '1', MemberSetLoadDistribution.LOAD_DISTRIBUTION_CONCENTRATED_2x2,
                        MemberSetLoadDirection.LOAD_DIRECTION_LOCAL_Z,
                        load_parameter=[False, False, False, force, distance_a, distance_b, distance_c])
    distance_a += increment
```

node_reactions:

```python
from RFEM.enums import *
from RFEM.Loads.membersetload import *
from RFEM.Results.resultTables import ResultTables

Model(False, "beams.rf6")

n_cases = 135
case_names = ["NL1-P" + str(i) for i in range(1, n_cases + 1)]

Output_Joints = [19, 45, 52, 61, 60, 67]
for j in Output_Joints:
    for i, name in enumerate(case_names):
        result_table = ResultTables.NodesSupportForces(CaseObjectType.E_OBJECT_TYPE_LOAD_CASE, i+1, 61)
        N = result_table[0]["support_force_p_z"]
        M = result_table[0]["support_moment_m_y"]
        print(name, N, M)```

**Comment by matchristmann** [comment]
Hello @BurakYyurt ,
thanks for your questions.

The perfomance issues stems from the update of the graphics inside RFEM 6. We are currently still working on increasing speed of our services.

The second provided script can be optimized by using a slightly different approach:

```python
from RFEM.enums import *
from RFEM.Loads.membersetload import *
from RFEM.Results.resultTables import ResultTables
from RFEM.Tools.GetObjectNumbersByType import GetObjectNumbersByType

import itertools

Model(False, "beams.rf6")

load_cases = GetObjectNumbersByType(ObjectTypes.E_OBJECT_TYPE_LOAD_CASE)
Output_Joints = [19, 45, 52, 61, 60, 67]

for j,i in itertools.product(Output_Joints, load_cases):
    result_table = ResultTables.NodesSupportForces(CaseObjectType.E_OBJECT_TYPE_LOAD_CASE, i, j)
    N = result_table[0]["support_force_p_z"]
    M = result_table[0]["support_moment_m_y"]
    print(Model.clientModel.service.get_load_case(i).name, N, M)

```

By using the function ``` GetObjectNumbersByType() ``` we can get the LoadCases in the model.
If you decide you want to export the support forces, have a look at our Example scripts on how to do so. `SteelDesignExcel.py` shows an easy way on how to get your data in Excel.

I hope this helps,
Matti

**Reply by BurakYyurt** [reply]
Hello Matti,

Thanks for the response. So basically, some operations, for the time being, are inherently slow. Which is a bit sad because I was really looking forward to get more involved. It will be a bit difficult for me to convince my supervisors to invest more time. 

I could not try "SteelDesingExcel.py" because there is a 12 member limit in the demo version. There seem to be no way to generate a model with less than 12 members
Incidentally, there is a webinar about this example next week. But those who don't have a Web Services license (which I believe most of the attendees) cannot run the code. It will be a bit weird to join the seminar but not be able to do the example.

**Comment by dogukankaratas** [comment, marked as answer]
Hi @BurakYyurt ,

I just made slight changes and improved the performance of your script. 

### Load Application
For the creation of load, please use Model class with True, with this you can avoid connection for every loop and therefore you'll gain some time. Also please add begin and finish modification this will help to improve connection time. 

Your script was taking 595 sec (which is not efficient at all), with the script below you can execute same operation in 18 sec.

```python
from RFEM.initModel import Model
from RFEM.LoadCasesAndCombinations.loadCase import LoadCase
from RFEM.Loads.membersetload import MemberSetLoad
from RFEM.enums import MemberSetLoadDistribution, MemberSetLoadDirection
import time

start = time.time()
Model(True, "beams.rf6")
Model.clientModel.service.begin_modification()

# Create Load Cases
n_cases = 135
case_names = ["NL1-P" + str(i) for i in range(1, n_cases + 1)]

increment = 0.100
distance_a = 0
distance_b = 0.620
distance_c = 1.480
force = 184000

for i, name in enumerate(case_names):
    LoadCase(i + 1, name, [False])
    MemberSetLoad.Force(1, i+1, '1', MemberSetLoadDistribution.LOAD_DISTRIBUTION_CONCENTRATED_2x2,
                        MemberSetLoadDirection.LOAD_DIRECTION_LOCAL_Z,
                        load_parameter=[False, False, False, force, distance_a, distance_b, distance_c])
    distance_a += increment
Model.clientModel.service.finish_modification()

end = time.time()
print(f"Total time is {end-start} seconds.")
```

### Reading Results
Also, same solution applies to reading results for the support nodes.  The code below took 32 seconds for me, which I believe acceptable for reading 810 result points. 

But there can be still a way to improve this which is in the ResultTables class. If you pass the object_no=0, it will read all of the nodes. I don't precisely know how you want to parse these results, therefore I couldn't create an example for it. Please try this, and let us know if you have any questions.

```python
from RFEM.enums import *
from RFEM.Loads.membersetload import *
from RFEM.Results.resultTables import ResultTables
import time

start = time.time()

Model(True, "beams.rf6")
Model.clientModel.service.begin_modification()

n_cases = 135
case_names = ["NL1-P" + str(i) for i in range(1, n_cases + 1)]

Output_Joints = [19, 45, 52, 61, 60, 67]
for j in Output_Joints:
    for i, name in enumerate(case_names):
        result_table = ResultTables.NodesSupportForces(CaseObjectType.E_OBJECT_TYPE_LOAD_CASE, i+1, 61)
        N = result_table[0]["support_force_p_z"]
        M = result_table[0]["support_moment_m_y"]
        print(name, N, M)

Model.clientModel.service.finish_modification()
end = time.time()

print(f"Total time is {end-start} seconds")
```

**Reply by BurakYyurt** [reply]
Hello @dogukankaratas

That looks just excellent. Changing False to True makes huge impact. I think I set the parameter to "False" because somewhere in the tutorials it should be False to edit an existing model. In a video or document or something, cant remember well. That's why I made it as such. Also, if the parameter is set to "False" then Model.clientModel.service.finish_modification() throws an error. So that's why I also removed them.

In any case, this was a great help now I understand better how things work.
Thanks!

**Reply by dogukankaratas** [reply]
Hi @BurakYyurt ,
I'm glad if this is helpful. Nevertheless, it's unacceptable that setting Model parameter false is slowing down the connection that much. We are going to work on it.
Thanks for the feedback.

**Comment by OndraMichal** [comment]
Hi @BurakYyurt,
setting `Model(False)` no longer slows down the process. Nevertheless, performance is under scrutiny and we try to improve it where possible.

**Comment by OndraMichal** [comment]
Using RFEM running in command line called [RFEM6Server](https://github.com/Dlubal-Software/RFEM_Python_Client/discussions/259) is also good option to speed things up. See our [wiki](https://github.com/Dlubal-Software/RFEM_Python_Client/wiki/RFEM6Server-RFEM-without-GUI)
